### PR TITLE
Set dynamic config values and default search attribute cache as disabled

### PR DESCRIPTION
**Pull request by cretz**
**What changed?**

* Added CLI and library support for setting individual dynamic config values
* Defaulted CLI (with opt-out) and test server to disable search attribute cache

**Why?**

For use with search attributes, the dev-only setting to disable cache is needed or there is a wait period before the search attributes are available for use.

This further cements that Temporalite is not for production use. This is the same config as dev docker compose at https://github.com/temporalio/docker-compose/blob/7c4f708c579cd735145e49816750ca4368ff5bf3/dynamicconfig/development-sql.yaml#L5. Any concerns here?

### CODE REVIEW

File: server.go
Comment (cretz) [inline server.go:101]: Technically I could build a fall back here, but there wasn't enough value (our NoopClient from before basically ensured nobody used this in their config anyways).

**Review by bergundy — APPROVED** [review]

**Review by robholland — APPROVED** [review]

File: cmd/temporalite/main.go
Comment (cretz) [inline cmd/temporalite/main.go:53]: Gonna probably remove this and instead set a default of true if they don't pass `--dynamic-config-value system.forceSearchAttributesCacheRefreshOnRead=false`.

File: cmd/temporalite/main_test.go
Comment (cretz) [inline cmd/temporalite/main_test.go:1]: Running `go run ./internal/copyright` puts a header that doesn't validate on `--verify-only`, but we can fix that later

File: cmd/temporalite/main.go
Comment (jlegrone) [inline cmd/temporalite/main.go:277]: This extra bit of explanation from your PR description would be helpful here too IMO.

```suggestion
				// If there is no config value for search attribute cache disabling,
				// default it to true. This is needed in order to avoid a wait period
				// before search attributes are available for use.
```

File: cmd/temporalite/main.go
Comment (jlegrone) [reply cmd/temporalite/main.go:282]: Also, instead of setting a default here, I think we should probably do it inside `temporalite.NewServer` so that library users get the same behavior.

File: cmd/temporalite/main.go
Comment (jlegrone) [inline cmd/temporalite/main.go:322]: Would there be a way to do any additional validation here eg. to check for typos in key names? It seems like keys are all consts in the dynamicconfig package, so I'm guessing there's not a good way to do this using reflection.

File: cmd/temporalite/main.go
Comment (jlegrone) [inline cmd/temporalite/main.go:155]: Should we add a `dynamic-config-file` flag as well? That would make it easier to share dynamic config profiles between multiple Temporalite (and regular Temporal server) instances.

I also wouldn't be mad if we added a shorthand `dcv` alias for this.

File: cmd/temporalite/main_test.go
Comment (jlegrone) [reply cmd/temporalite/main_test.go:1]: Well, there's the problem 🤦 https://github.com/temporalio/temporalite/blob/f3d60547a26df3fcde912d22d56f8aaedbecff69/internal/copyright/main.go#L171

I can try whipping up a quick fix so we can merge with the right copyright headers.

File: cmd/temporalite/main.go
Comment (jlegrone) [inline cmd/temporalite/main.go:323]: Constraints sound nice... what are they and is there future work we should track to support them?

File: cmd/temporalite/main_test.go
Comment (jlegrone) [inline cmd/temporalite/main_test.go:43]: Should we error out instead when the same key is passed multiple times, or is there a legit use case or convenience factor in favor of this?

File: options.go
Comment (jlegrone) [inline options.go:141]: Commented above, but should we make this the default and instead add an option for enabling the cache? We could comment to the effect of changing this default if/when we are able to remove the waiting period after registering search attributes, and if you need stable behavior use `WithDynamicConfigValue`.

File: cmd/temporalite/main_test.go
Comment (jlegrone) [reply cmd/temporalite/main_test.go:1]: Just kidding that's always been a useless function. But it's a simple fix: https://github.com/temporalio/temporalite/pull/137

File: cmd/temporalite/main.go
Comment (cretz) [reply cmd/temporalite/main.go:282]: I originally had the same concerns but thought Temporalite was going for "development defaults" even in the CLI. I would be happy only setting this in `temporaltest`. Let me confirm with @bergundy who wanted this defauted in the CLI.

File: cmd/temporalite/main.go
Comment (cretz) [reply cmd/temporalite/main.go:155]: The dynamic config file can be set in the config file itself. We need to make the config file part more robust rather than accepting dynamic config directly (same for all sorts of other config options, of which dynamic config file is just one).

File: cmd/temporalite/main.go
Comment (cretz) [reply cmd/temporalite/main.go:323]: I don't really know, I'd have to speak to the server team. My use case and most use cases I've seen don't have these constraints. We can block this PR for that if we really need to.

File: options.go
Comment (cretz) [reply options.go:141]: I think it should be opt-in, like persistence-disabled, especially now that we are going to turn it off by default on the CLI.

File: cmd/temporalite/main.go
Comment (bergundy) [reply cmd/temporalite/main.go:282]: This makes me think that having the `temporalite "yolo"` command for common test setup (includes `--ephemeral` too) is starting to make sense.

Comment (jlegrone) [reply cmd/temporalite/main.go:282]: IMO Temporalite values go something like this

1. User experience (for development use cases)
2. Security
3. Performance
4. Simplicity
5. Configurability

So I'm very happy to trade off some performance for a better development experience. We can always add a `--production` flag to swap around a few defaults and keep things simple.

Another thought is that we could add flags to register search attributes at startup like we do for namespaces, and essentially have the best of both worlds since we'd be able to ensure that each of those SAs is registered synchronously at startup time, without disabling caching completely.

File: cmd/temporalite/main_test.go
Comment (jlegrone) [reply cmd/temporalite/main_test.go:1]: Please go ahead and resolve this thread after updating the header now that the fix is merged.

File: cmd/temporalite/main.go
Comment (bergundy) [reply cmd/temporalite/main.go:282]: > Another thought is that we could add flags to register search attributes at startup like we do for namespaces. This would essentially be the best of both worlds since we'd be able to ensure that each of those SAs is registered synchronously, without disabling caching completely.

I'd rather users do this with the SDK so their tests can run against any server implementation without customization.

**Comment by jlegrone** [comment]
This lgtm if we turn off the SA cache in `temporaltest` only for now. For the record I'm open to making this the default behavior from the CLI in the future, but would love to see a benchmark comparing perf.

@cretz would you mind adding a note on dynamic config to the relevant section of the README as well?

File: cmd/temporalite/main.go
Comment (cretz) [reply cmd/temporalite/main.go]: Removed default from CLI

File: cmd/temporalite/main.go
Comment (cretz) [reply cmd/temporalite/main.go]: No longer applicable, this is no longer in CLI

**Review by jlegrone — APPROVED** [review]

**Comment by cretz** [comment]
I think some schedule-to-close is too aggressive for Windows in some tests causing flakes.

**Pull request merged**
